[TA1290]  Identifying the replicas that can connect to target

mynktl opened the pull request:
Changes : 
Added support for identifying replicas connecting to the target.
With this change,
1. The target will allow session from any replica until the number of connected(active or closed session) replicas reaches to replication factor limit. Once the replication factor limit reaches, Target will not allow any new replica to connect.
2. If the target has already an active session with a replica(`zvol_guid` is `12`) and same replica(with `zvol_guid` `12`) tries to make a new session with the target then the target will disconnect from the previous session and allow the new replica to connect.

Bug fix for replica cleanup path in data connection (in `handle_data_conn_error`).

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

Commit:
[TA1290] Identifying the replicas that can connect to target

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

Commit:
[TA1699] Error handling for mutiple responses for a mgmt command from a replica
         and some minor fixes for rebuild and memory leak

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

Commit:
- Setting zvol_guid to replica port instead of a static value in istgt
- Fixing minor issue in closing replica connection in data interface

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

satbirchhikara commented on src/data_conn.c:
```diff
@@ -293,6 +293,8 @@ handle_data_conn_error(replica_t *r)
 
 	MTX_LOCK(&r->r_mtx);
 	if (epoll_ctl(r->epollfd, EPOLL_CTL_DEL, r->iofd, NULL) == -1) {
+		shutdown(r->iofd, SHUT_RDWR);
```
If epoll_ctl system call got an error, it implies that something weird happened in the system. In such case better to get rid-off this replica. Why we do by returning from here ?

mynktl replied:
yes. This is replica removal path from the target, So we can ignore error here.

satbirchhikara commented on src/replication.c:
```diff
@@ -488,6 +497,57 @@ trigger_rebuild(spec_t *spec)
+		kr->zvol_guid = new_replica->zvol_guid;
+		kr->is_connected = true;
+		ret = true;
+		TAILQ_INSERT_TAIL(&spec->identified_replica, kr, next);
```
I can see that we are inserting replica in to "identified_replica" list but we are not removing anything from this list anywhere in code. Is it by design or we missed freeing it ?

mynktl replied:
yes. It is by design. With this, we are allowing first `replication_factor` number of replicas to connect.

satbirchhikara commented on src/replication.c:
```diff
@@ -1893,23 +1978,25 @@ empty_mgmt_q_of_replica(replica_t *r)
+					    mgmt_cmd->io_hdr, NULL, mgmt_cmd);
+					break;
+				case ZVOL_OPCODE_STATS:
+					update_spec_stats(r->spec, mgmt_cmd->io_hdr,
```
How about rename to "handle_update_spec_stats" just for symmetry ??

mynktl replied:
ok

satbirchhikara commented on src/replication.c:
```diff
@@ -488,6 +497,57 @@ trigger_rebuild(spec_t *spec)
+	if (!found && connected_replica < spec->replication_factor) {
+		kr = malloc(sizeof (known_replica_t));
+		kr->zvol_guid = new_replica->zvol_guid;
+		kr->is_connected = true;
```
What if replica is already there in identified_replica list but not connected(is_connected == false) where would we set it to true when we move this replica to ready_queue (spec->rq) ? Connected_replica should be count of replicas which are actually connected i.e. kr->is_connected should be true for those replicas otherwise we are just counting stale replica.

mynktl replied:
Yes.`is_connected` should be set to `true` if replica re-connects.

Commit:
Addressing review comments.

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

mynktl reviewed (commented):
Thanks for reviewing changes @satbirchhikara Sir.
I have tried to address all review comments, please have a look and let me know if any changes are required.

satbirchhikara commented on src/replication.c:
```diff
@@ -381,13 +384,17 @@ start_rebuild(void *buf, replica_t *replica, uint64_t data_len)
 	mgmt_cmd->io_bytes = 0;
 	mgmt_cmd->data = buf;
 	mgmt_cmd->mgmt_cmd_state = WRITE_IO_SEND_HDR;
+	mgmt_cmd->rcomm_mgmt = NULL;
```
Since we are zeroing this memory, we do not need code at line 394 and at 397.

mynktl replied:
ok

satbirchhikara commented on src/replication.c:
```diff
@@ -933,6 +1003,7 @@ send_replica_handshake_query(replica_t *replica, spec_t *spec)
 	int ret = 0;
 
 	mgmt_cmd = malloc(sizeof(mgmt_cmd_t));
+	memset(mgmt_cmd, 0, sizeof (mgmt_cmd_t));
```
Once we zero out this memory, we do not need code at line 1016 and 1019. We can remove those lines.

mynktl replied:
ok

satbirchhikara commented on src/replication.c:
```diff
@@ -998,6 +1069,7 @@ send_replica_snapshot(spec_t *spec, replica_t *replica, uint64_t io_seq,
 	int ret = 0;
 
 	mgmt_cmd = malloc(sizeof(mgmt_cmd_t));
+	memset(mgmt_cmd, 0, sizeof (mgmt_cmd_t));
```
Same comment goes here too.

mynktl replied:
ok

satbirchhikara commented on src/replication.c:
```diff
@@ -1230,6 +1302,7 @@ send_replica_query(replica_t *replica, spec_t *spec, zvol_op_code_t opcode)
 	mgmt_cmd_t *mgmt_cmd;
 
 	mgmt_cmd = malloc(sizeof(mgmt_cmd_t));
+	memset(mgmt_cmd, 0, sizeof (mgmt_cmd_t));
```
Same goes here.

mynktl replied:
ok

satbirchhikara reviewed (approved):
Some minor comments given. Please have a look and let me know if you need more clarification.

Commit:
Addressing review comments

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

mynktl committed:
Merge branch 'replication' into TA1290

Commit:
- Assertion fix for istgt_integration reported in travis build
- Bug fix for replica cleanup path in data connection (in handle_data_conn_error).

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

vishnuitta commented on src/data_conn.c:
```diff
@@ -292,12 +301,7 @@ handle_data_conn_error(replica_t *r)
 	MTX_UNLOCK(&spec->rq_mtx);
 
 	MTX_LOCK(&r->r_mtx);
-	if (epoll_ctl(r->epollfd, EPOLL_CTL_DEL, r->iofd, NULL) == -1) {
```
why this change?

mynktl replied:
Here, `handle_data_conn_error` was returning an error if `epoll_ctl` fails which may block management thread which will be waiting for `replica_thread`.
Since it is a cleanup path, we can ignore `epoll_ctl` return value and continue further cleanup.

vishnuitta commented on src/replication.c:
```diff
@@ -500,6 +499,59 @@ trigger_rebuild(spec_t *spec)
+{
+	known_replica_t *kr = NULL;
+	int familiar_replicas = 0;
+	bool ret = false, found = false;
```
its better to change the name of 'ret' to 'newly_connected'

mynktl replied:
ok

vishnuitta commented on src/replication.c:
```diff
@@ -500,6 +499,59 @@ trigger_rebuild(spec_t *spec)
+				MTX_UNLOCK(&spec->rq_mtx);
+				REPLICA_LOG("Disconnecting replica(%lu)'s old "
+				    "session\n", old_replica->zvol_guid);
+				handle_mgmt_conn_error(old_replica, 0, NULL, 0);
```
during this, what happens for any pending events received in 'init_replication' thread?

mynktl replied:
Good find!
We will not allow a new connection due to complexity. At this point, we won't be able to know any pending events on old replica session.

vishnuitta commented on src/replication.c:
```diff
@@ -500,6 +499,59 @@ trigger_rebuild(spec_t *spec)
+	 * then the target will disconnect from the previous session and
+	 * allow a new session
+	 */
+	if (found && !ret) {
```
let us continue the old one, and error out the new connection due to complexities in looking for any pending IOs / pending epoll signals.
Is there any reason why we should consider new connection?

mynktl replied:
let us continue the old one, and error out the new connection due to complexities in looking for any pending IOs / pending EPOLL signals.
>Good find! We can avoid new connection if we set keepalive for TCP socket. I have created an issue(**US2827**) for the same.

Is there any reason why we should consider a new connection?
>If the system(replica node) crash happens then target won't get any event regarding replica disconnect. Once system/replica comes up then it will try to make a new connection. Now, the target will have two sessions from the same replica.
As of now, the old session will take some time for disconnect event since we are not setting keepalive for a socket. Due to this, we are considering the new connection.

vishnuitta commented on src/replication.c:
```diff
@@ -873,13 +925,21 @@ update_replica_entry(spec_t *spec, replica_t *replica, int iofd)
-	if(is_rf_replicas_connected(spec, replica)) {
+	if (is_rf_replicas_connected(spec, replica)) {
 		MTX_UNLOCK(&spec->rq_mtx);
 		ISTGT_ERRLOG("failed to verify replica(ip:%s port:%d "
```
seems like, this log need to be fixed (not related to this PR)

mynktl replied:
ok

vishnuitta commented on src/replication.c:
```diff
@@ -873,13 +925,21 @@ update_replica_entry(spec_t *spec, replica_t *replica, int iofd)
 		goto replica_error;
 	}
+
+	if (!is_replica_familiar(spec, replica)) {
```
'is_replica_new_or_familar' may be good?

mynktl replied:
ok

Commit:
Addressing review comments

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

vishnuitta commented on src/replication.c:
```diff
@@ -500,6 +499,54 @@ trigger_rebuild(spec_t *spec)
+		REPLICA_ERRLOG("replica(%lu) has already an active session\n",
+		    kr->zvol_guid);
+		/*
+		 * handle_mgmt_conn_error will update identified_replica status
```
thats a very good explanation @mynktl ..
I'm thinking whether same scenario (i.e., another connection from same replica) can apply to other error scenarios in update_replica_entry after setting zvol_guid.
We need to move this check for is_replica_new_or_familiar right after setting zvol_guid in update_replica_entry.
Sorry to ask one more change along with above thing: can you change is_replica_new_or_familiar to is_replica_newly_connected and with some comments saying 'This function returns whether connection is newly created from a familiar replica or newly connected from an unknown replica that target is waiting for'?

mynktl replied:
ok

vishnuitta commented on src/replication.c:
```diff
@@ -2617,6 +2676,7 @@ initialize_volume(spec_t *spec, int replication_factor, int consistency_factor)
 	TAILQ_INIT(&spec->rcommon_waitq);
 	TAILQ_INIT(&spec->rq);
 	TAILQ_INIT(&spec->rwaitq);
+	TAILQ_INIT(&spec->identified_replica);
```
should we destroy the list during destroy_volume?

mynktl replied:
PR which contains destroy_volume function is not yet merged. I thought of handling destroying identified_replica in that PR #54 once this PR gets merged.

Commit:
Addressing review comments

Signed-off-by: mayank <mayank.patel@cloudbyte.com>

vishnuitta commented on src/replication.c:
```diff
@@ -2365,6 +2434,13 @@ handle_mgmt_conn_error(replica_t *r, int sfd, struct epoll_event *events, int ev
+		}
+	}
 	MTX_UNLOCK(&r->spec->rq_mtx);
 
```
we can add an ASSERT for debug build if kr is NULL

vishnuitta replied:
ok.. may be, not required..
if too many replicas gets connected, there are chances that kr may be NULL

vishnuitta reviewed (approved):
changes are good

vishnuitta committed:
Merge branch 'replication' into TA1290
